assets/model_monitoring/components/src/action_analyzer/contracts/action_detectors/low_retreival_score_index_action_detector.py

```diff
@@ -3,49 +3,143 @@
+LOW_RETRIEVAL_SCORE_QUERY_RATIO_THRESHOLD = 0.1
+
+
+class LowRetrievalScoreIndexActionDetector(ActionDetector):
     """Low retrieval score index action detector class."""
 
     def __init__(self,
                  index_id: str,
                  violated_metrics: list[str],
-                 llm_summary_enabled: str) -> None:
+                 query_intention_enabled: str) -> None:
         """Create a low retrieval score index action detector.
 
         Args:
             index_id(str): the index asset id.
             violated_metrics(List[str]): violated e2e metrics
-            llm_summary_enabled(str): enable llm generated summary. Accepted values: true or false.
+            query_intention_enabled(str): enable llm generated summary. Accepted values: true or false.
+            max_positive_sample_size(int): (Optional) max positive sample size in the action.
         """
         self.index_id = index_id
         self.violated_metrics = violated_metrics
-        super().__init__(llm_summary_enabled)
+        super().__init__(query_intention_enabled)
 
     def preprocess_data(self, df: pandas.DataFrame) -> pandas.DataFrame:
         """Preprocess the data for action detector.
 
+            1. check if all violated metrics are available. If not, call evaluate sdk to get the e2e metrics.
+            2. extract extra fields from the root span for action.
+            3. convert the dataframe from trace level to span level.
+
         Args:
             df(pandas.DataFrame): input pandas dataframe.
 
         Returns:
             pandas.DataFrame: preprocessed pandas dataframe.
         """
-        pass
+        missed_metrics = get_missed_metrics(self.violated_metrics, df.columns.tolist())
+
+        if missed_metrics != []:
+            df = calculate_e2e_metrics(df, missed_metrics)
 
-    def detect(self, df: pandas.DataFrame, llm_client: LLMClient) -> list(Action):
+        return extract_fields_from_debugging_info(df, self.index_id)
+
+    def detect(self, df: pandas.DataFrame, llm_client: LLMClient, aml_deployment_id=None) -> list[Action]:
         """Detect the action.
 
         Args:
             df(pandas.DataFrame): input pandas dataframe.
             llm_client(LLMClient): LLM client used to get some llm scores/info for action.
+            aml_deployment_id(str): (Optional) aml deployment id for the action.
 
         Returns:
-            list(Action): list of actions.
+            list[Action]: list of actions.
         """
-        pass
+        # get llm retrieval score
+        df[INDEX_SCORE_LLM_COLUMN] = df.apply(get_retrieval_score, axis=1, args=(llm_client,))
+        df = df[df[INDEX_SCORE_LLM_COLUMN] != DEFAULT_LLM_SCORE]
+
+        action_list = []
+        for metric in self.violated_metrics:
+            low_retrieval_score_df = df[(df[metric] < METRICS_VIOLATION_THRESHOLD) &
+                                        (df[INDEX_SCORE_LLM_COLUMN] < METRICS_VIOLATION_THRESHOLD)]
```

Comment:
in theory these two thresholds (metrics violation threshold can be different, currently we can keep the same value

Reply:
added thresholds for retrieval score